pandas/core/reshape/merge.py

```diff
@@ -731,7 +732,11 @@ def _maybe_add_join_keys(self, result, left_indexer, right_indexer):
                 if mask.all():
                     key_col = rvals
                 else:
-                    key_col = Index(lvals).where(~mask, rvals)
+                    # Might need to be IntIndex not Index
+                    if isinstance(lvals, SparseArray):
+                        key_col = Index(lvals.get_values()).where(~mask, rvals)
```

Comment:
I'm not sure if this has memory or performance issues. But this is the best solution I could come to with this. The other solution would be to look at using lvals.sp_index and implement a where on it that works.

One thing I have noticed is that IntIndex doesn't act quite like Index which makes for doing these masks tricky in sparse land.

Reply:
It'd be nice to avoid `get_values` if possible.